cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/WrappedStageOutputFormat.java

```diff
@@ -0,0 +1,108 @@
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.OutputCommitter;
+import org.apache.hadoop.mapreduce.OutputFormat;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+
+/**
+ * A delegating output format that catches exceptions thrown during execution of a call
+ * and wraps them in a {@link WrappedStageException}.
+ * This class is primarily used to associate the exception with a specific stage name in a pipeline,
+ * helping in better debugging and error tracking.
+ *
+ * <p>
+ * The class delegates the actual calling operation to another {@link TrackingOutputFormat} instance
+ * and ensures that any exceptions thrown are caught and rethrown as a {@link WrappedStageException}
+ * , which includes the stage name where the error occurred.
+ * </p>
+ *
+ * @param <K> type of key to write
+ * @param <V> type of value to write
+ */
+public class WrappedStageOutputFormat<K, V> extends OutputFormat<K, V> {
+  public static final String WRAPPED_CLASS_NAME = "io.cdap.pipeline.wrapped.output.classname";
+  public static final String WRAPPED_STAGE_NAME = "io.cdap.pipeline.wrapped.stage.name";
+
+  @Override
+  public RecordWriter<K, V> getRecordWriter (TaskAttemptContext context) {
+    try {
+      return new WrappedStageOutputRecordWriter<>(
+        getDelegate(context.getConfiguration()).getRecordWriter(context),
+        getStageName(context.getConfiguration()));
+    } catch (Exception e) {
+      throw new WrappedStageException(e, getStageName(context.getConfiguration()));
+    }
+  }
+
+  @Override
+  public void checkOutputSpecs (JobContext jobContext) {
+    try {
+      getDelegate(jobContext.getConfiguration()).checkOutputSpecs(jobContext);
+    } catch (Exception e) {
+      throw new WrappedStageException(e, getStageName(jobContext.getConfiguration()));
+    }
+  }
+
+  @Override
+  public OutputCommitter getOutputCommitter (TaskAttemptContext context) {
+    try {
+      return new WrappedStageOutputCommitter(
+        getDelegate(context.getConfiguration()).getOutputCommitter(context),
+        getStageName(context.getConfiguration()));
+    } catch (Exception e) {
+      throw new WrappedStageException(e, getStageName(context.getConfiguration()));
+    }
+  }
+
+  /**
+   * Returns the delegating {@link OutputFormat} based on the configuration.
+   *
+   * @param conf the Hadoop {@link Configuration} for this output format
+   */
+  private OutputFormat<K, V> getDelegate (Configuration conf) {
+    String delegateClassName = conf.get(WRAPPED_CLASS_NAME);
+    try {
+      //noinspection unchecked
+      OutputFormat<K, V> outputFormat = (OutputFormat<K, V>) conf.getClassLoader()
+        .loadClass(delegateClassName)
+        .newInstance();
+      if (outputFormat instanceof Configurable) {
+        ((Configurable) outputFormat).setConf(conf);
+      }
+      return outputFormat;
```

Comment:
I think we can directly wrap this in the TrackingOutputFormat instead of having the extra layer of delegation. For example, add a constructor to TrackingOutputFormat that takes the delegate directly, in which case it doesn't do the dynamic class instantiation.

Then maybe rename the class to StageTrackingOutputFormat.

Reply:
ack, done.